static/js/timerender.ts

```diff
@@ -24,6 +24,80 @@ export function clear_for_testing(): void {
     next_timerender_id = 0;
 }
 
+type DateFormat = "short" | "medium" | "day" | "full";
+type DateWithTimeFormat = "shortWithTime" | "mediumWithTime" | "longWithTime";
+type TimeFormat = "time";
+
+type DateOrTimeFormat = DateFormat | TimeFormat | DateWithTimeFormat;
+
+// The function takes in an optional type parameter and returns
+// an options object that can be used to format a date/time using the Intl.DateTimeFormat API.
+// This function allows easy and consistent formatting of dates and times
+// with different options to customize the format according to the specified type
+function get_format_options_for_type(type: DateOrTimeFormat): Intl.DateTimeFormatOptions {
+    const isHourCycleTwentyFourHour = user_settings.twenty_four_hour_time;
+
+    const timeFormatOptions: Intl.DateTimeFormatOptions = isHourCycleTwentyFourHour
+        ? {hourCycle: "h23", hour: "2-digit", minute: "2-digit"}
+        : {
+              hourCycle: "h12",
+              hour: "numeric",
+              minute: "2-digit",
+          };
+
+    const dayFormatOptions: Intl.DateTimeFormatOptions = {weekday: "long"};
+    const fullFormatOptions: Intl.DateTimeFormatOptions = {dateStyle: "full"};
+
+    const shortFormatOptions: Intl.DateTimeFormatOptions = {day: "numeric", month: "short"};
+    const mediumFormatOptions: Intl.DateTimeFormatOptions = {
+        ...shortFormatOptions,
+        year: "numeric",
+    };
+    const longFormatOptions: Intl.DateTimeFormatOptions = {
+        ...mediumFormatOptions,
+        weekday: "short",
+    };
+
+    switch (type) {
+        case "day": // Wednesday
+            return dayFormatOptions;
+        case "full": // Wednesday, July 27, 2016
+            return fullFormatOptions;
+        case "time": // 01:30 PM
+            return timeFormatOptions;
+        case "short": // Jul 27
+            return shortFormatOptions;
+        case "shortWithTime": // Jul 27, 01:30 PM
+            return {...shortFormatOptions, ...timeFormatOptions};
+        case "medium": // Jul 27, 2016
+            return mediumFormatOptions;
+        case "mediumWithTime": // Jul 27, 2016, 01:30 PM
+            return {...mediumFormatOptions, ...timeFormatOptions};
+        case "longWithTime": // Wed, Jul 27, 2016, 13:30
+            return {...longFormatOptions, ...timeFormatOptions};
```

Comment:
Hmm, thinking about this more, I think we can give these sensible names where someone seeing code using this can know what the format means roughly without going and looking it up in this function:

* day => weekday
* full => weekday_dayofyear_year
* short => dayofyear
* shortWithTime => dayofyear_time
* medium => dayofyear_year
* mediumWithTime => dayofyear_year_time
* longWithTime => weekday_dayofyear_year_time

Reply:
added to the list:
time => hour_min

and the new one used in `get_full_datetime` function:
hour_min_sec